### DISCUSSION TITLE
How to save/load a fingerprint to/from a string?

### DISCUSSION THREAD

**Discussion by R1ncewind** [rdkit/rdkit#3874] [Q&A] [answered]
I have a database containing chemicals with smiles strings as a property. Now I would like to go through all the chemicals and add a fingerprint as an additional property so I can later on rapidly compute distances between the molecules. So I need to export the fingerprint as a string, write that to the db and later query that property, load said string into a new BitVector and compare it with another BitVector.
I thought that could be possible with `rdkit.DataStructs.cDataStructs.ExplicitBitVect.FromBase64` and `ToBase64`, but
`FromBase64` throws the following error at me:
```
ArgumentError: Python argument types in
    ExplicitBitVect.FromBase64()
did not match C++ signature:
    FromBase64(ExplicitBitVect {lvalue}, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
```
Sadly, I could not find any documentation on either of the functions or `ExplicitBitVect` in the docs to help me.
Can somebody point me in the right direction?

**Comment by ptosco** [comment, marked as answer]
@R1ncewind 

It is not a static method, you first need to call the `ExplicitBitVect` constructor with the number of bits your original fingerprint had, and then initialize the new `ExplicitBitVect` from your `base64` encoded string:
```
from rdkit.Chem import AllChem
from rdkit.DataStructs.cDataStructs import ExplicitBitVect

mol = Chem.MolFromSmiles('c1ccccn1')
fp1 = AllChem.GetMorganFingerprintAsBitVect(mol, 2, 2048)

base64_fp = fp1.ToBase64()

base64_fp
'4P///wAIAAAJAAAA6QMICl4pCIT9BBkCDLkA'

fp_from_base64 = ExplicitBitVect(2048)
fp_from_base64.FromBase64(base64_fp)

fp_from_base64.ToBase64()
'4P///wAIAAAJAAAA6QMICl4pCIT9BBkCDLkA'

fp_from_base64.ToBase64() == base64_fp
True
```

**Reply by R1ncewind** [reply]
Thank you for your reply.
Is there any documentation where I could have read up on this? 

I could not find `ExplicitBitVect`, `GetMorganFingerprintAsBitVect` and others in the [python documentation](https://www.rdkit.org/docs/api-docs.html), but their methods seem to differ from what is described in the C++ docs.
For example `ExplicitBitVect` does not have a `toString()` method in python that pickles it, but has `toBase64()` which is not in the C++ doc.

To find out about the methods of `ExplicitBitVect` I had to use
```
object_methods = [method_name for method_name in dir(vec_1)
                  if callable(getattr(vec_1, method_name))]
print(object_methods)
```
and there has to be something more serviceable, right?

**Comment by ptosco** [comment]
They are in the Python docs:

`GetMorganFingerprintAsBitVect`:
http://rdkit.org/docs/source/rdkit.Chem.rdMolDescriptors.html?highlight=getmorganfingerprintasbitvect#rdkit.Chem.rdMolDescriptors.GetMorganFingerprintAsBitVect

`ExplicitBitVect`:
http://rdkit.org/docs/source/rdkit.DataStructs.cDataStructs.html?highlight=explicitbitvect#rdkit.DataStructs.cDataStructs.ExplicitBitVect

**Comment by mgarard** [comment]
Easy way:
To save:
You can use the Vernalis "Fingerprint to string" node or use the Column Expressions node to cast the fingerprint to a string.

To retrieve:
Vernalis node called "fingerprint from binary string"

I had a large file and had to loop by line so my solution was with a python node to save with the commands:
from rdkit import Chem
from rdkit.Chem import AllChem as ac
from rdkit.Chem import DataStructs # for ToBitString()

fp = AllChem.GetMorganFingerprintAsBitVect(Chem.MolFromSmiles( line.split()[0] ),2,nBits=1024)
newline = line.strip('\n') + "\t" + fp.ToBitString() 

Then to get it out  There is a Vernalis node called "fingerprint from binary string" which converts it to a bitvector type again.
I should have mentioned Vernalis nodes are in KNIME which is free to use.

Otherwise, python would be:
bitvect_fingerprint = DataStructs.CreateFromBitString( "100101001110001" )
